### PR TITLE
AHI Cloud Detection based on IR Cloud Mask

### PR DESCRIPTION
**Pull request by xdmjolly**
TYPE: new feature

KEYWORDS: WRFDA, AHI, Cloud detection

SOURCE: Deqin Li (Liaoning Meteorological Bureau of CMA/NCAR), Dongmei Xu (NUIST/NCAR),  Jonathan Guerrette (NCAR).

DESCRIPTION OF CHANGES:
A cloud detection method is added for AHI radiance data assimilation based on an infrared-only cloud mask (CM) algorithm.

LIST OF MODIFIED FILES: 
	A:   var/da/da_physics/da_trop_wmo.inc
	A:   var/da/da_radiance/mod_clddet_geoir.f90
	M:   Registry/registry.var
	M:   var/build/da.make
	M:   var/build/depend.txt
	M:   var/da/da_define_structures/da_define_structures.f90
	M:   var/da/da_physics/da_physics.f90
	M:   var/da/da_radiance/da_allocate_rad_iv.inc
	M:   var/da/da_radiance/da_crtm.f90
	M:   var/da/da_radiance/da_deallocate_radiance.inc
	M:   var/da/da_radiance/da_get_innov_vector_crtm.inc
	M:   var/da/da_radiance/da_initialize_rad_iv.inc
        M:   var/da/da_radiance/da_qc_ahi.inc
	M:   var/da/da_radiance/da_radiance.f90
	M:   var/da/da_radiance/da_radiance1.f90
	M:   var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc
	M:   var/da/da_radiance/da_setup_radiance_structures.inc

TESTS CONDUCTED: 
WRFDA regression tests ran successfully with gnu/ifort+mpt on Cheyenne.

RELEASE NOTE: New capability for an IR-only cloud detection scheme for AHI radiance data assimilation.
Wu et al., 2020, Improving forecasts of a record-breaking rainstorm in Guangzhou by assimilating every 10-min AHI radiances with WRF 4DVAR. Atmospheric Research, 239, 104912.

Zhuge, X., Zou, X., 2016. Test of a modified infrared-only ABI cloud mask algorithm for
AHI radiance observations. J. Appl. Meteorol. Climatol. 55 (11), 2529–2546.

### CODE REVIEW

**Comment by davegill** [comment]
@xdmjolly 
Please add the text of the jenkins email in one of these lower comments

File: var/da/da_radiance/da_qc_ahi.inc [hunk 1]
```diff
@@ -137,105 +419,68 @@ subroutine da_qc_ahi (it, i, nchan, ob, iv)
...
                 if (iv%instid(i)%info%proc_domain(1,n)) &
                         nrej_omb_abs(k) = nrej_omb_abs(k) + 1
             end if			
          end do ! nchan
-
-         ! 1. Cloud detection scheme MMR in Auligné (2014).or. PF in Xu et al., (2016)
-         !---------------------------------------------
-         if ((use_clddet==1 .or. use_clddet==2) .and. (.not.use_satcv(2))) then
-            iv%instid(i)%cloud_flag(:,n) = qc_good
-
```
Comment (liujake) [inline var/da/da_radiance/da_qc_ahi.inc]: why you delete the existing cloud detection schemes? Can we keep them?

File: var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc [hunk 2]
```diff
@@ -1,6 +1,6 @@
-subroutine da_read_obs_netcdf4ahi_jaxa (iv, infile_tb, infile_clp)
+subroutine da_read_obs_netcdf4ahi_jaxa (iv, infile_tb, infile_tbp)
    !--------------------------------------------------------
-   !  Purpose: read in JAXA AHI Level-1 and Level-2 data in NETCDF4 format
+   !  Purpose: read in JAXA AHI Level-1 and Level-1 (previous) data in NETCDF4 format
    !           and form innovation structure
```
Comment (liujake) [inline var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc]: Unclear what is the difference between 'Level-1' and 'Level-1 (previous)'. Please make it more understandable.

File: var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc [hunk 3]
```diff
@@ -10,14 +10,15 @@ subroutine da_read_obs_netcdf4ahi_jaxa (iv, infile_tb, infile_clp)
    !                  and deallocate sequential data structure
    !
    !  HISTORY: 2016/10/22 - Creation         Yuanbing Wang, NUIST/CAS, NCAR/NESL/MMM/DAS
+   !  HISTORY: 2020/03/01 - Add clear sky cloud detection procedures   Dongmei Xu, NUIST/CAS, NCAR/NESL/MMM/DAS  
    !  To be devoloped: 1.time information; 2.dimension sequence
```
Comment (liujake) [inline var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc]: Same comment about affiliation here.

File: var/da/da_radiance/mod_qc.f90 [hunk 4]
```diff
@@ -0,0 +1,338 @@
+  module mod_qc
+!   use netcdf
```
Comment (liujake) [inline var/da/da_radiance/mod_qc.f90]: I would use a more relevant name like mod_clddet_ahi or sth more general mod_clddet_geoir if it is applicable to both AHI and ABI.

**Comment by Plantain** [comment]
This commit appears to depend on a static file "static_ahi.nc" in da_read_obs_netcdf4ahi_jaxa.inc
@xdmjolly @liujake where can this file be found?

It also looks like this commit breaks using JAXA NetCDF4's for usecases that don't use use_clddet_zz, see cases of this popping up in the forums like https://forum.mmm.ucar.edu/phpBB3/viewtopic.php?f=18&t=9512&p=20935#p20935
https://forum.mmm.ucar.edu/phpBB3/viewtopic.php?t=10206

**Comment by liujake** [comment]
@Plantain Somehow this cloud detection scheme was omitted in WRFDA 4.2 release note. It is now added in https://www2.mmm.ucar.edu/wrf/users/wrfda/updates-4.2.html. The download link for static_ahi.bc is provided.

**Comment by Plantain** [comment]
@liujake Thank you!

I would again draw your attention though to this change breaking any use of JAXA Himawari8 NetCDF's with anything other than clddet_zz though, due to the changes in https://github.com/wrf-model/WRF/pull/1139/files#diff-4a0f1b1feda554b73ca5fa2f34c70168af0e721800b791a6d66f4fe0dffd274bR237 and da_read_obs_netcdf4ahi_jaxa.inc.

Currently if clddet_zz = 0, WRFDA will crash.

I don't think that was an intended part of this change!

File: var/da/da_define_structures/da_define_structures.f90 [hunk 5]
```diff
@@ -548,6 +553,7 @@ module da_define_structures
       real,    pointer     :: satazi(:) 
       real,    pointer     :: solzen(:) 
       real,    pointer     :: solazi(:) 
+	  real,    pointer     :: tropt(:)
```
[the remark this answers is not in the document]
Comment (fzrfzr) [reply var/da/da_define_structures/da_define_structures.f90]: I'm encountering difficulties in assimilating sunflower data. I've adopted the cloud detection method proposed by Zhuge and Zou, but I'm unable to obtain the "tropt" (tropopause temperature) data, which is preventing me from completing the drawing of the cloud detection map. Does anyone know how this data is obtained?

**Pull request merged**